### PR TITLE
Redux exercise

### PR DESCRIPTION
**Pull request by JoaquinBattilana**
## Summary

-Used redux for communication between components in the webpage

## Screenshots

redux_exercise

## Trello Card

https://trello.com/c/gc1L7UoX/13-redux-i-actions-reducers-store

### CODE REVIEW

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/Search/index.js [hunk 1]
```diff
@@ -1,20 +1,32 @@
...
-  render() {
-    return <input className={styles.input} type="text" placeholder="Search by name" onChange={this.search} />;
-  }
-}
+const Search = ({ onSearch }) => (
+  <input
+    className={styles.input}
+    type="text"
+    placeholder="Search by name"
+    onChange={e => {
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/Search/index.js]: you can remove braces here `e => onSearch(..`

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/Search/index.js [hunk 2]
```diff
@@ -1,20 +1,32 @@
...
+  />
+);
 
 Search.propTypes = {
   onSearch: func.isRequired
 };
 
-export default Search;
+const mapDispatchToProps = dispatch => ({
+  onSearch(value) {
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/Search/index.js]: you can also declare this as 
```
onSearch: value => dispatch(bookActionsCreators.searchBook(value))
```
which i personally prefer 😛 

also id change this to ` e => dispatch(bookActions.searchBook(e.target.value.toLowerCase()))`
and in the component simply do `onClick={onSearch}`

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 3]
```diff
@@ -1,32 +1,29 @@
...
+import cartActionsCreators from '@redux/cart/actions';
 
 import styles from './styles.scss';
 
-class Item extends PureComponent {
-  addItem = () => {
-    const { item, addItem } = this.props;
-    addItem(item.id);
+class Item extends Component {
+  handlePlusClick = () => {
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js]: no braces needed here

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 4]
```diff
@@ -1,32 +1,29 @@
...
+    this.props.addItem(this.props.item.id);
   };
-
-  removeItem = () => {
-    const { item, removeItem } = this.props;
-    removeItem(item.id);
+  handleTrashClick = () => {
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js]: nor here

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 5]
```diff
@@ -1,32 +1,29 @@
...
+    this.props.removeItem(this.props.item.id);
   };
-
   render() {
-    const { item } = this.props;
     return (
       <li className={styles.item}>
-        <h3 className={styles.title}>{item.name}</h3>
+        <h3 className={styles.title}>{this.props.item.name}</h3>
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js]: you can destructure this before the return to make it easier to read like
```
const { item: { name, quantity } } = this.props;
```

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 6]
```diff
@@ -41,4 +38,16 @@ Item.propTypes = {
   removeItem: func.isRequired
 };
 
-export default Item;
+const mapDispatchToProps = dispatch => ({
+  addItem(itemId) {
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js]: id also declare this as 
```
addItem: itemId => ...,
removeItem: itemId => ...
```

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 7]
```diff
@@ -1,50 +1,55 @@
...
-    return <Item key={item.id} item={item} addItem={addItem} removeItem={removeItem} />;
-  };
+  renderItem = item => <Item key={item.id} item={item} />;
 
   render() {
-    const { data } = this.props;
     return (
       <Fragment>
-        <Button className={styles.buttonCart} onClick={this.toggleContent}>
+        <Button className={styles.buttonCart} onClick={this.props.toggleContent}>
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: i'd also destructure `this.props` for this `render`

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 8]
```diff
@@ -1,50 +1,55 @@
...
 };
 
-export default ShoppingCart;
+const mapStateToProps = ({ cart }) => ({
+  bookSelected: cart.bookSelected,
+  open: cart.open
+});
+
+const mapDispatchToProps = dispatch => ({
+  toggleContent() {
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: `toggleContent: () => dispatch...`

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js [hunk 9]
```diff
@@ -57,22 +40,50 @@ class App extends Component {
...
 
-export default App;
+const mapStateToProps = ({ books, cart }) => ({
+  books: books.books,
+  bookSelected: cart.bookSelected,
+  originalData: books.originalData
+});
+
+const mapDispatchToProps = dispatch => ({
+  getOriginalBooks() {
```
Comment (gon-escalante) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js]: `getOriginalBooks: () => ..` and same for the others

**Review by gon-escalante — CHANGES_REQUESTED** [review]

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/Search/index.js [hunk 10]
```diff
@@ -1,20 +1,23 @@
...
-  search = e => {
-    this.props.onSearch(e.target.value.toLowerCase());
-  };
-
-  render() {
-    return <input className={styles.input} type="text" placeholder="Search by name" onChange={this.search} />;
-  }
-}
+const Search = ({ onSearch }) => (
+  <input className={styles.input} type="text" placeholder="Search by name" onChange={onSearch} />
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/Search/index.js]: Extract "Search by name" to a constant file

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 11]
```diff
@@ -1,32 +1,28 @@
...
-  };
-
-  removeItem = () => {
-    const { item, removeItem } = this.props;
-    removeItem(item.id);
-  };
-
+class Item extends Component {
+  handlePlusClick = () => this.props.addItem(this.props.item.id);
+  handleTrashClick = () => this.props.removeItem(this.props.item.id);
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js:13]: Add a space after and before of this function

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js [hunk 12]
```diff
@@ -57,22 +40,44 @@ class App extends Component {
...
           ) : (
             <div className={styles.noData}>
               <h2 className={styles.title}>No Data</h2>
             </div>
           )}
         </div>
-        {this.state.bookSelected.length ? (
-          <ShoppingCart data={this.state.bookSelected} addItem={this.addItem} removeItem={this.removeItem} />
-        ) : null}
+        {this.props.bookSelected.length ? <ShoppingCart /> : null}
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js]: Extract the keys of props and then you don't need to repeat this.props all the time

File: react/finger-exercises/redux/02 - react-redux/src/redux/cart/reducer.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+// i don't like the sintax of this func but is the only way that the linter don't bother
+const incrementById = (arr, id) =>
+  arr.map(elem => {
+    if (elem.id === id) {
+      const aux = { ...elem };
+      aux.quantity += 1;
+      return aux;
+    }
+    return elem;
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/redux/cart/reducer.js]: What do you think to write it like arr.map(elem => elem.id === id ? {...elem, quantity: elem.quantity + 1 } : elem)?

**Review by erikankodelia — CHANGES_REQUESTED** [review]

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js (hunk 11 above)
Comment (JoaquinBattilana) [reply react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js:13]: linter don't let me do this, it appear as an error

File: react/finger-exercises/redux/02 - react-redux/src/redux/cart/reducer.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+const initialState = {
+  bookSelected: [],
+  open: false
+};
+
+// i don't like the sintax of this func but is the only way that the linter don't bother
+const incrementById = (arr, id) =>
+  arr.map(elem => (elem.id === id ? { ...elem, quantity: elem.quantity + 1 } : elem));
+function reducer(state = initialState, action) {
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/redux/cart/reducer.js:12]: Add a space before this function

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js [hunk 15]
```diff
@@ -1,78 +1,84 @@
...
           ) : (
             <div className={styles.noData}>
               <h2 className={styles.title}>No Data</h2>
             </div>
           )}
         </div>
-        {this.state.bookSelected.length ? (
-          <ShoppingCart data={this.state.bookSelected} addItem={this.addItem} removeItem={this.removeItem} />
-        ) : null}
+        {bookSelected.length ? <ShoppingCart /> : null}
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js]: You can write bookSelected.length && <ShoppingCart />

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js (hunk 11 above)
Comment (JoaquinBattilana) [reply react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js:13]: Already fixed this, i don't know why it doesn't appear outdated.

**Review by erikankodelia — APPROVED** [review]

**Review by gon-escalante — APPROVED** [review]

**Pull request merged**